<!DOCTYPE html><html lang="en" ><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="generator" content="Jekyll v4.2.2" /><meta property="og:title" content="The most naive phylogenetic reconstruction algorithm" /><meta property="og:locale" content="en" /><meta name="description" content="The full source code for this post can be found on GitHub." /><meta property="og:description" content="The full source code for this post can be found on GitHub." /><link rel="canonical" href="/posts/possibly-the-most-naive-phylogenetic-reconstruction-algorithm/" /><meta property="og:url" content="/posts/possibly-the-most-naive-phylogenetic-reconstruction-algorithm/" /><meta property="og:site_name" content="Guilherme Sena" /><meta property="og:type" content="article" /><meta property="article:published_time" content="2022-07-18T00:00:00+00:00" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="The most naive phylogenetic reconstruction algorithm" /><meta name="twitter:site" content="@senacompbio" /><meta name="google-site-verification" content="google23fa0f88983501d9.html" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"BlogPosting","dateModified":"2022-07-25T02:56:05+00:00","datePublished":"2022-07-18T00:00:00+00:00","description":"The full source code for this post can be found on GitHub.","headline":"The most naive phylogenetic reconstruction algorithm","mainEntityOfPage":{"@type":"WebPage","@id":"/posts/possibly-the-most-naive-phylogenetic-reconstruction-algorithm/"},"url":"/posts/possibly-the-most-naive-phylogenetic-reconstruction-algorithm/"}</script><title>The most naive phylogenetic reconstruction algorithm | Guilherme Sena</title><link rel="apple-touch-icon" sizes="180x180" href="/assets/img/favicons/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="/assets/img/favicons/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="/assets/img/favicons/favicon-16x16.png"><link rel="manifest" href="/assets/img/favicons/site.webmanifest"><link rel="shortcut icon" href="/assets/img/favicons/favicon.ico"><meta name="apple-mobile-web-app-title" content="Guilherme Sena"><meta name="application-name" content="Guilherme Sena"><meta name="msapplication-TileColor" content="#da532c"><meta name="msapplication-config" content="/assets/img/favicons/browserconfig.xml"><meta name="theme-color" content="#ffffff"><link rel="preconnect" href="https://fonts.googleapis.com" ><link rel="dns-prefetch" href="https://fonts.googleapis.com" ><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin><link rel="dns-prefetch" href="https://fonts.gstatic.com" crossorigin><link rel="preconnect" href="https://fonts.googleapis.com" ><link rel="dns-prefetch" href="https://fonts.googleapis.com" ><link rel="preconnect" href="https://cdn.jsdelivr.net" ><link rel="dns-prefetch" href="https://cdn.jsdelivr.net" ><link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Lato&family=Source+Sans+Pro:wght@400;600;700;900&display=swap"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4/dist/css/bootstrap.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.11.2/css/all.min.css"><link rel="stylesheet" href="/assets/css/style.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@1.0.1/dist/bootstrap-toc.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/magnific-popup@1/dist/magnific-popup.min.css"> <script src="https://cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js"></script> <script type="text/javascript"> class ModeToggle { static get MODE_KEY() { return "mode"; } static get MODE_ATTR() { return "data-mode"; } static get DARK_MODE() { return "dark"; } static get LIGHT_MODE() { return "light"; } static get ID() { return "mode-toggle"; } constructor() { if (this.hasMode) { if (this.isDarkMode) { if (!this.isSysDarkPrefer) { this.setDark(); } } else { if (this.isSysDarkPrefer) { this.setLight(); } } } let self = this; /* always follow the system prefers */ this.sysDarkPrefers.addEventListener("change", () => { if (self.hasMode) { if (self.isDarkMode) { if (!self.isSysDarkPrefer) { self.setDark(); } } else { if (self.isSysDarkPrefer) { self.setLight(); } } self.clearMode(); } self.notify(); }); } /* constructor() */ get sysDarkPrefers() { return window.matchMedia("(prefers-color-scheme: dark)"); } get isSysDarkPrefer() { return this.sysDarkPrefers.matches; } get isDarkMode() { return this.mode === ModeToggle.DARK_MODE; } get isLightMode() { return this.mode === ModeToggle.LIGHT_MODE; } get hasMode() { return this.mode != null; } get mode() { return sessionStorage.getItem(ModeToggle.MODE_KEY); } /* get the current mode on screen */ get modeStatus() { if (this.isDarkMode || (!this.hasMode && this.isSysDarkPrefer)) { return ModeToggle.DARK_MODE; } else { return ModeToggle.LIGHT_MODE; } } setDark() { $('html').attr(ModeToggle.MODE_ATTR, ModeToggle.DARK_MODE); sessionStorage.setItem(ModeToggle.MODE_KEY, ModeToggle.DARK_MODE); } setLight() { $('html').attr(ModeToggle.MODE_ATTR, ModeToggle.LIGHT_MODE); sessionStorage.setItem(ModeToggle.MODE_KEY, ModeToggle.LIGHT_MODE); } clearMode() { $('html').removeAttr(ModeToggle.MODE_ATTR); sessionStorage.removeItem(ModeToggle.MODE_KEY); } /* Notify another plugins that the theme mode has changed */ notify() { window.postMessage({ direction: ModeToggle.ID, message: this.modeStatus }, "*"); } } /* ModeToggle */ const toggle = new ModeToggle(); function flipMode() { if (toggle.hasMode) { if (toggle.isSysDarkPrefer) { if (toggle.isLightMode) { toggle.clearMode(); } else { toggle.setLight(); } } else { if (toggle.isDarkMode) { toggle.clearMode(); } else { toggle.setDark(); } } } else { if (toggle.isSysDarkPrefer) { toggle.setLight(); } else { toggle.setDark(); } } toggle.notify(); } /* flipMode() */ </script><body data-spy="scroll" data-target="#toc" data-topbar-visible="true"><div id="sidebar" class="d-flex flex-column align-items-end"><div class="profile-wrapper text-center"><div id="avatar"> <a href="/" class="mx-auto"> <img src="/assets/img/favicons/android-chrome-512x512.png" alt="avatar" onerror="this.style.display='none'"> </a></div><div class="site-title mt-3"> <a href="/">Guilherme Sena</a></div><div class="site-subtitle font-italic">Create. Not for money, fame or recognition, but for the pure joy of creating something and sharing it.</div></div><ul class="w-100"><li class="nav-item"> <a href="/" class="nav-link"> <i class="fa-fw fas fa-home ml-xl-3 mr-xl-3 unloaded"></i> <span>HOME</span> </a><li class="nav-item"> <a href="/categories/" class="nav-link"> <i class="fa-fw fas fa-stream ml-xl-3 mr-xl-3 unloaded"></i> <span>CATEGORIES</span> </a><li class="nav-item"> <a href="/archives/" class="nav-link"> <i class="fa-fw fas fa-archive ml-xl-3 mr-xl-3 unloaded"></i> <span>ARCHIVES</span> </a><li class="nav-item"> <a href="/about/" class="nav-link"> <i class="fa-fw fas fa-info-circle ml-xl-3 mr-xl-3 unloaded"></i> <span>ABOUT</span> </a></ul><div class="sidebar-bottom mt-auto d-flex flex-wrap justify-content-center align-items-center"> <button class="mode-toggle btn" aria-label="Switch Mode"> <i class="fas fa-adjust"></i> </button> <span class="icon-border"></span> <a href="https://github.com/guilhermesena1" aria-label="github" target="_blank" rel="noopener"> <i class="fab fa-github"></i> </a> <a href="https://twitter.com/senacompbio" aria-label="twitter" target="_blank" rel="noopener"> <i class="fab fa-twitter"></i> </a> <a href=" javascript:location.href = 'mailto:' + ['desenabr','usc.edu'].join('@')" aria-label="email" > <i class="fas fa-envelope"></i> </a></div></div><div id="topbar-wrapper"><div id="topbar" class="container d-flex align-items-center justify-content-between h-100 pl-3 pr-3 pl-md-4 pr-md-4"> <span id="breadcrumb"> <span> <a href="/"> Home </a> </span> <span>The most naive phylogenetic reconstruction algorithm</span> </span> <i id="sidebar-trigger" class="fas fa-bars fa-fw"></i><div id="topbar-title"> Post</div><i id="search-trigger" class="fas fa-search fa-fw"></i> <span id="search-wrapper" class="align-items-center"> <i class="fas fa-search fa-fw"></i> <input class="form-control" id="search-input" type="search" aria-label="search" autocomplete="off" placeholder="Search..."> </span> <span id="search-cancel" >Cancel</span></div></div><div id="main-wrapper" class="d-flex justify-content-center"><div id="main" class="container pl-xl-4 pr-xl-4"><div class="row"><div id="core-wrapper" class="col-12 col-lg-11 col-xl-9 pr-xl-4"><div class="post pl-1 pr-1 pl-md-2 pr-md-2"><h1 data-toc-skip>The most naive phylogenetic reconstruction algorithm</h1><div class="post-meta text-muted"> <span> Posted <em class="" data-ts="1658102400" data-df="ll" data-toggle="tooltip" data-placement="bottom"> Jul 18, 2022 </em> </span> <span> Updated <em class="" data-ts="1658717765" data-df="ll" data-toggle="tooltip" data-placement="bottom"> Jul 24, 2022 </em> </span><div class="d-flex justify-content-between"> <span> By <em> <a href="https://twitter.com/senacompbio">Guilherme Sena</a> </em> </span><div> <span class="readtime" data-toggle="tooltip" data-placement="bottom" title="3515 words"> <em>19 min</em> read</span></div></div></div><div class="post-content"><p>The full source code for this post can be found on <a href="https://github.com/guilhermesena1/phylogenetic-by-genome">GitHub</a>.</p><p>This post was motivated by the following question: <em>what would be the simplest problem we could formulate to introduce a computer science student to the field of bioinformatics?</em> This is the answer I came up with, and is a (hopefully) fun attempt to analyze easily accessible biological data with very little code or effort. The question we will address is the following:</p><p><strong>given the genome sequences of a set of species, can we reconstruct their speciation history?</strong></p><p>Below we will provide a brief introduction to the problem, then formulate it more formally and solve it with some code, after which we will discuss the code’s correctness, efficiency and potential limitations.</p><h1 id="introduction">Introduction</h1><p>Two species that are phenotypically very similar are expected to have had a recent common ancestor. In other words, not too long ago (which in “evolution time” may mean hundreds of thousands of years ago), a speciation event occurred, and the two resulting species evolved independently, resulting in the observable differences. The longer ago the speciation event, the more different species are today.</p><p>In theory we can (and have) reconstruct speciation events by observing physical traits (phenotypes) of a pair of species. For example, we do not need to look at DNA to have a strong conviction that the Chinese hamster and the mouse are close siblings evolutionarily. Just look at them! (and take a second to appreciate how adorable they are)</p><p><img data-src="https://i.ibb.co/L5SGKd5/post-1.png" alt="a Chinese hamster (left) and a mouse (right)" data-proofer-ignore></p><p>Conversely, relying purely on traits can be misleading. Bats and birds have wings and other similar physical traits, but bats are more closely related to cats than to birds. Similarly, mammalian aquatic animals like whales and dolphins are closer to wolves and alpacas than to fish (by “closer” I mean that the speciation event of the most recent common ancestor occurred more recently). These are examples of <a href="https://en.wikipedia.org/wiki/Convergent_evolution">convergent evolution</a>, where two species have similar traits not necessarily because they are related by recent common ancestry, but because they are under similar selective pressure to acquire those traits.</p><p>With recent genome sequencing technologies, we can objectively measure similarity between two species: simply measure the “similarity” between the genome sequences, where the sequences are strings of characters (A, C, G, T) representing chromosomes. This is easier said than done. First, we need to define what “similarity” of sequences means. When speciation occurs, independent genome divergence occurs through a sequence of transformations: mutations, nucleotide insertions, deletions, duplications of certain sequences, and possible translocations and inversions of large portions of the genome. Simply put, pieces of DNA can move around, mutate, copy-paste themselves, and other small operations that can significantly change the resulting sequence. Can you think of a similarity metric between two strings that would identify similar genomic sequences under these potential change operations? Think if edit distance would be enough. Yes? No? To me at least it is not immediately obvious. Simple translocations of genome sequences can significantly change the edit distance.</p><p>In most bioinformatics methods used today, comparison of sequences is done through (possibly some modification of) the <a href="https://www.sciencedirect.com/science/article/abs/pii/0022283681900875">local alignment of two sequences</a>. Given sequences $A$ and $B$, of sizes $|A|$ and $|B|$, the local alignment of $A$ and $B$ is the most similar substring of $A$ to a substring of $B$. Objectively, we devise a scoring scheme that rewards similar letters in $A$ and penalize letter differences, insertions and deletions. The local alignment can be found through the Smith-Waterman algorithm in $O(|A| |B|)$ by filling a $|A| \times |B|$ matrix. The local alignment score $S[i, j]$ of the first $i$ characters of $A$ and the first $j$ characters of $B$ can be found by the following dynamic programming iteration: $S[i, 0] = S[0, j] = 0$ and for $i, j &gt; 0$:</p>\[S[i, j] = \begin{cases} S[i - 1, j - 1] + s(A[i], B[j]) \\ S[i - 1], j] - \delta \\ S[i, j - 1] - \delta \\ 0 \end{cases}\]<p>where $s(x, y) = 1$ if $x = y$ or $-\mu$ if $x \neq y$. Here $\mu$ is the penalty of mismatch of two letters, and $\delta$ is the penalty to insert or delete a character in either $A$ or $B$.</p><p>This is not a post about local alignments, so this extremely simple definition of local alignment will suffice for our purposes. We can even assume $\mu = \delta = 1$, which is possibly the simplest alignment scheme. I should, however point out that the resulting score matrix $S$ contains a lot of information about the evolutionary relationship between $A$ and $B$. The highest-scoring location in $S$ is the most similar subsequence of the two, but other high-scoring cells in the matrix can potentially identify subsequences that translocated across the genome and diverged independently. Possible inversions can be found by aligning $A$ to the reverse-complement of $B$ and vice-versa. By studying the alignment matrix, we can learn, in great detail, what likely happened to certain genome subsequences, and quantify how exactly they diverged between the two species.</p><p><strong>The limitation of alignments</strong>: Quantifying similarities between species through local alignments is not immediately obvious. The highest alignment score (which, recall, is the most similar subsequence) may not be representative of the global similarity, and this becomes more true when comparing more distant species. We can use other high scores to reconstruct the divergence of subsequences, but it is not entirely obvious how to combine high scores in the matrix into one global metric of similarity between sequences. Furthermore, many genomes, including most mammalians, are billions of sequences long. A $O(|A| |B|)$ algorithm is prohibitive both in time and memory.</p><p>My proposition to solve this problem is to take a step back and think about the absolutely most naive way to compare two sequences: Consider small sequences of length $k$ (for example, $k = 12$), which we will call $k$-mers. What if we just counted $k$-mers in two sequences and calculated the sum of squares of their differences? This is analogous to measuring the similarity of two books based on their word frequencies. Two dictionaries written by different people would be very similar, and so would two math textbooks that would constantly use words like “function”, “variable”, “number”&lt; etc.</p><p>Note: we are shifting our thought process from “similarity” -a number that is higher the more similar two sequences are - to “distance”. A distance function between two strings is a function $d$ such that, for strings $A$ and $B$, $d(A, B) = 0$ if and only if $A = B$, $d(A,B) = d(B,A)$ and for any three strings $A, B, C$, $d(A,B) \leq d(A,C) + d(B,C)$. The edit distance is an example of a distance metric between strings, and we can also prove that our $k$-mer distance also is, since our $k$-mer counts induces a $4^k$-dimensional vector for which Euclidean geometry properties apply.</p><p>This approach is not very rigorous, and it may not necessarily work, but it is a start when thinking of a complex problem. Take a second to think about how this could break: Could two highly similar sequences have highly different $k$-mer counts? Conversely, could two very different sequences have similar $k$-mer counts? If you cannot convince yourself that it can break easily, then the approach may have some merit.</p><p>In fact, hopefully I will be able to show you that this extremely naive approach leads to somewhat satisfiable results. The take-home message should be that these simple approaches should often not be overlooked.</p><p>By the way, this is the most basic example of an <a href="https://academic.oup.com/bib/article-abstract/15/3/343/182355">alignment-free genome comparison method</a>. Alignment-free sequence comparison research a beautiful field in and of itself with various applications in metagenomics, virology and population genetics.</p><h1 id="problem-formulation">Problem formulation</h1><p>We are given a set of $n$ species and their reference genome sequences $S_1, \cdots, S_n$, which is the concatenation of all of their chromosomes. We wish to obtain the following.</p><ul><li>A $n \times n$ distance matrix between all pairs of species<li>A phylogenetic tree reconstructing the ancestral speciation events that led to the observed species, based on the similarity matrix.</ul><p>For this post, we will use the following species: human, gorilla, mouse, Chinese hamster, cat, alpaca and whale. Quick pause for some more adorable pictures.</p><p><img data-src="https://i.ibb.co/Rcwck5C/datasets.png" alt="our adorable species set" data-proofer-ignore></p><p>Genomes of each species are given as <a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/">FASTA files</a>. Those are human-readable text files, where lines indicating chromosome names start with the <code class="language-plaintext highlighter-rouge">&gt;</code> character (e.g. <code class="language-plaintext highlighter-rouge">&gt;chr1</code>), and below each such line we see the genome sequence, with some line breaks after at most 80 characters (so the sequence can be read in a terminal). For example, below are the first lines of the cat genome (<code class="language-plaintext highlighter-rouge">felCat9.fa</code>):</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
</pre><td class="rouge-code"><pre><span class="nv">$ </span><span class="nb">head </span>genomes/felCat9.fa
<span class="o">&gt;</span>chrA1
atcaggagatctagatgcctggagaggagtggagaaaacgggaaaccctc
ttATGggaagaggtaatatgtatttctccttcgaatataaaaaaagtaaa
aagaaggaaaacttaccaaattcacttatgagccattcattaccctgata
ccaaaaccagataaagccctccactaaaaccaaaactgcagcggcgcctt
gtgggctcggtcggttttactgtccaactcttaatttcagattaggaaat
aatcttgcggtgcatgggttcaagtcccacgttggaccctgccatgacag
tgtggggaatggctaggattctctctctccctgtctctctgcccctccct
cacttttttgtactctaaggaaagaaataaacatttaaaaaaatgttgaa
aattttttaaataaaactgcataccaatagccttgatgagtatgtatgcc
</pre></table></code></div></div><p>For our problem, we will download some FASTA files, inflate them, count the k-mers and save the k-mer frequencies into a table, which will then be turned into a distance matrix and a phylogenetic tree in R.</p><h1 id="obtaining-the-data">Obtaining the data</h1><p>Put this in a text file and call it <code class="language-plaintext highlighter-rouge">genome_urls.txt</code>. These are the URLs for the FASTA files of our species.</p><div class="language-plaintext highlighter-rouge"><div class="code-header"> <span data-label-text="Plaintext"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
</pre><td class="rouge-code"><pre>$ cat genome_urls.txt
https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
https://hgdownload.soe.ucsc.edu/goldenPath/gorGor6/bigZips/gorGor6.fa.gz
https://hgdownload.soe.ucsc.edu/goldenPath/mm39/bigZips/mm39.fa.gz
https://hgdownload.soe.ucsc.edu/goldenPath/criGriChoV2/bigZips/criGriChoV2.fa.gz
https://hgdownload.soe.ucsc.edu/goldenPath/felCat9/bigZips/felCat9.fa.gz
https://hgdownload.soe.ucsc.edu/goldenPath/vicPac2/bigZips/vicPac2.fa.gz
https://hgdownload.soe.ucsc.edu/goldenPath/balAcu1/bigZips/balAcu1.fa.gz
</pre></table></code></div></div><p>To download everything, on the same directory of your <code class="language-plaintext highlighter-rouge">genome_urls.txt</code>, use <code class="language-plaintext highlighter-rouge">wget</code> on each line of the file. Then create a <code class="language-plaintext highlighter-rouge">genomes</code> directory, unzip all the <code class="language-plaintext highlighter-rouge">.fa</code> files and move them to the <code class="language-plaintext highlighter-rouge">genomes</code> folder as shown below.</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
</pre><td class="rouge-code"><pre><span class="k">for </span>i <span class="k">in</span> <span class="si">$(</span><span class="nb">cat </span>genome_urls.txt<span class="si">)</span><span class="p">;</span> <span class="k">do </span><span class="nb">echo</span> <span class="nv">$i</span><span class="p">;</span> wget <span class="k">${</span><span class="nv">i</span><span class="k">}</span><span class="p">;</span> <span class="k">done
</span><span class="nb">gunzip</span> <span class="k">*</span>.fa.gz<span class="p">;</span>
<span class="nb">mkdir </span>genomes
<span class="nb">mv</span> <span class="k">*</span>.fa genomes
</pre></table></code></div></div><p>Now that we have the FASTA files, we will read them in C++ and create a $k$-mer frequency matrix.</p><h1 id="the-c-code">The C++ code</h1><p>Let us first write a struct that will store $k$-mer frequencies for a genome. We will use $k = 12$, and pre-allocate counts for all $4^{12}$ possible $k$-mers. We will assume counts fit into a 32-bit integer, since $2^{32}$ is larger than any of the genomes we are using.</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
</pre><td class="rouge-code"><pre><span class="k">struct</span> <span class="nc">KmerStats</span> <span class="p">{</span>
  <span class="n">KmerStats</span><span class="p">()</span> <span class="p">{</span> <span class="n">kmer_count</span> <span class="o">=</span> <span class="n">vector</span><span class="o">&lt;</span><span class="kt">uint32_t</span><span class="o">&gt;</span><span class="p">(</span><span class="n">num_kmers</span><span class="p">,</span> <span class="mi">0</span><span class="p">);</span> <span class="p">}</span>
  <span class="kt">void</span> <span class="n">count_kmers</span><span class="p">(</span><span class="k">const</span> <span class="n">string</span> <span class="o">&amp;</span><span class="n">chrom</span><span class="p">);</span>
  <span class="n">vector</span><span class="o">&lt;</span><span class="kt">uint32_t</span><span class="o">&gt;</span> <span class="n">kmer_count</span><span class="p">;</span>
  <span class="k">static</span> <span class="k">const</span> <span class="kt">uint32_t</span> <span class="n">kmer_size</span> <span class="o">=</span> <span class="mi">12</span><span class="p">;</span>
  <span class="k">static</span> <span class="k">const</span> <span class="kt">uint32_t</span> <span class="n">num_kmers</span> <span class="o">=</span> <span class="p">(</span><span class="mi">2</span> <span class="o">&lt;&lt;</span> <span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">kmer_size</span><span class="p">));</span>
<span class="p">};</span>
</pre></table></code></div></div><p>The function <code class="language-plaintext highlighter-rouge">count_kmers</code> takes a chromosome and increments the vector <code class="language-plaintext highlighter-rouge">kmer_count</code>. We will encode $k$-mers as 24-bit numbers, with 2 bits per letter. Here we have <code class="language-plaintext highlighter-rouge">00 = A</code>, <code class="language-plaintext highlighter-rouge">01 = C</code>, <code class="language-plaintext highlighter-rouge">10 = G</code>, and <code class="language-plaintext highlighter-rouge">11 = T</code>, so the number <code class="language-plaintext highlighter-rouge">0b110011001010010100011011</code> is the sequence <code class="language-plaintext highlighter-rouge">TATAGGCCACGT</code>.</p><p>Here is the implementation of <code class="language-plaintext highlighter-rouge">count_kmers</code>:</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
</pre><td class="rouge-code"><pre><span class="kt">void</span>
<span class="n">KmerStats</span><span class="o">::</span><span class="n">count_kmers</span><span class="p">(</span><span class="k">const</span> <span class="n">string</span> <span class="o">&amp;</span><span class="n">chrom</span><span class="p">)</span> <span class="p">{</span>
  <span class="kt">uint32_t</span> <span class="n">mer</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span>

  <span class="k">auto</span> <span class="n">itr</span><span class="p">(</span><span class="n">begin</span><span class="p">(</span><span class="n">chrom</span><span class="p">));</span>
  <span class="k">const</span> <span class="k">auto</span> <span class="n">lim</span><span class="p">(</span><span class="n">end</span><span class="p">(</span><span class="n">chrom</span><span class="p">));</span>
  <span class="k">for</span> <span class="p">(</span><span class="kt">size_t</span> <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">itr</span> <span class="o">!=</span> <span class="n">lim</span> <span class="o">&amp;&amp;</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="n">kmer_size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">;</span> <span class="o">++</span><span class="n">i</span><span class="p">,</span> <span class="o">++</span><span class="n">itr</span><span class="p">)</span>
    <span class="n">shift_hash_key</span><span class="p">(</span><span class="o">*</span><span class="n">itr</span><span class="p">,</span> <span class="n">mer</span><span class="p">);</span>

  <span class="k">for</span> <span class="p">(;</span> <span class="n">itr</span> <span class="o">!=</span> <span class="n">lim</span><span class="p">;</span> <span class="o">++</span><span class="n">itr</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">shift_hash_key</span><span class="p">(</span><span class="o">*</span><span class="n">itr</span><span class="p">,</span> <span class="n">mer</span><span class="p">);</span>
    <span class="o">++</span><span class="n">kmer_count</span><span class="p">[</span><span class="n">mer</span><span class="p">];</span>
  <span class="p">}</span>
<span class="p">}</span>
</pre></table></code></div></div><p>The function <code class="language-plaintext highlighter-rouge">shift_hash_key</code> simply shifts a k-mer two bits, appends a new letter to the end of it, then discards the first number by applying an AND (<code class="language-plaintext highlighter-rouge">&amp;</code>) operation to the number <code class="language-plaintext highlighter-rouge">0b111111111111111111111111</code> (twenty-four 1s), which is <code class="language-plaintext highlighter-rouge">1 &lt;&lt; (2*KmerStats::kmer_size) - 1</code>.</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
</pre><td class="rouge-code"><pre><span class="kr">inline</span> <span class="kt">void</span>
<span class="nf">shift_hash_key</span><span class="p">(</span><span class="k">const</span> <span class="kt">uint8_t</span> <span class="n">c</span><span class="p">,</span> <span class="kt">uint32_t</span> <span class="o">&amp;</span><span class="n">mer</span><span class="p">)</span> <span class="p">{</span>
  <span class="k">static</span> <span class="k">const</span> <span class="kt">uint32_t</span> <span class="n">hash_mask</span> <span class="o">=</span> <span class="n">KmerStats</span><span class="o">::</span><span class="n">num_kmers</span> <span class="o">-</span> <span class="mi">1</span><span class="p">;</span>
  <span class="n">mer</span> <span class="o">=</span> <span class="p">((</span><span class="n">mer</span> <span class="o">&lt;&lt;</span> <span class="mi">2</span><span class="p">)</span> <span class="o">|</span> <span class="n">encode_char</span><span class="p">[</span><span class="n">c</span><span class="p">])</span> <span class="o">&amp;</span> <span class="n">hash_mask</span><span class="p">;</span>
<span class="p">}</span>
</pre></table></code></div></div><p>Finally, we write a function to take a file name, open it, read the chromosomes and process them onto a <code class="language-plaintext highlighter-rouge">KmerStats</code> object. Here we use a pre-allocation size of 250 million for the <code class="language-plaintext highlighter-rouge">chrom</code> string, which is larger than the largest chromosome we expect. This allows us to re-use memory each time we read a chromosome.</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
</pre><td class="rouge-code"><pre><span class="kt">void</span>
<span class="nf">process_species</span><span class="p">(</span><span class="k">const</span> <span class="n">string</span> <span class="o">&amp;</span><span class="n">file</span><span class="p">,</span> <span class="n">KmerStats</span> <span class="o">&amp;</span><span class="n">v</span><span class="p">)</span> <span class="p">{</span>
  <span class="c1">// get file size</span>
  <span class="n">ifstream</span> <span class="n">in</span><span class="p">(</span><span class="n">file</span><span class="p">);</span>

  <span class="k">if</span> <span class="p">(</span><span class="o">!</span><span class="n">in</span><span class="p">)</span>
    <span class="k">throw</span> <span class="n">runtime_error</span><span class="p">(</span><span class="s">"cannot open file "</span> <span class="o">+</span> <span class="n">file</span><span class="p">);</span>

  <span class="k">static</span> <span class="k">const</span> <span class="kt">size_t</span> <span class="n">RESERVE_SIZE</span> <span class="o">=</span> <span class="mi">250000000</span><span class="p">;</span>
  <span class="n">string</span> <span class="n">chrom</span><span class="p">;</span>
  <span class="n">chrom</span><span class="p">.</span><span class="n">reserve</span><span class="p">(</span><span class="n">RESERVE_SIZE</span><span class="p">);</span>

  <span class="n">string</span> <span class="n">line</span><span class="p">;</span>
  <span class="k">while</span> <span class="p">(</span><span class="n">getline</span><span class="p">(</span><span class="n">in</span><span class="p">,</span> <span class="n">line</span><span class="p">))</span> <span class="p">{</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="sc">'&gt;'</span><span class="p">)</span>
      <span class="n">copy</span><span class="p">(</span><span class="n">begin</span><span class="p">(</span><span class="n">line</span><span class="p">),</span> <span class="n">end</span><span class="p">(</span><span class="n">line</span><span class="p">),</span> <span class="n">back_inserter</span><span class="p">(</span><span class="n">chrom</span><span class="p">));</span>
    <span class="k">else</span>
      <span class="n">process_chrom</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">v</span><span class="p">);</span>
  <span class="p">}</span>

  <span class="n">process_chrom</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">v</span><span class="p">);</span> <span class="c1">// last chromosome after EOF</span>
  <span class="n">in</span><span class="p">.</span><span class="n">close</span><span class="p">();</span>
<span class="p">}</span>
</pre></table></code></div></div><p>The <code class="language-plaintext highlighter-rouge">process_chrom</code> function simply encodes the string in two bits per letter, replacing any non-ACGT character to a random character among the two. This ensures that, when counting $k$-mers, we only see ACGTs in our chromosome.</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
</pre><td class="rouge-code"><pre><span class="kr">inline</span> <span class="kt">void</span>
<span class="nf">process_chrom</span><span class="p">(</span><span class="n">string</span> <span class="o">&amp;</span><span class="n">chrom</span><span class="p">,</span> <span class="n">KmerStats</span> <span class="o">&amp;</span><span class="n">v</span><span class="p">)</span> <span class="p">{</span>
  <span class="c1">// makes sure letters are only ACGT:</span>
  <span class="k">for</span> <span class="p">(</span><span class="k">auto</span> <span class="n">it</span><span class="p">(</span><span class="n">begin</span><span class="p">(</span><span class="n">chrom</span><span class="p">));</span> <span class="n">it</span> <span class="o">!=</span> <span class="n">end</span><span class="p">(</span><span class="n">chrom</span><span class="p">);</span> <span class="o">++</span><span class="n">it</span><span class="p">)</span>
    <span class="o">*</span><span class="n">it</span> <span class="o">=</span> <span class="p">((</span><span class="n">encode_char</span><span class="p">[</span><span class="k">static_cast</span><span class="o">&lt;</span><span class="kt">uint8_t</span><span class="o">&gt;</span><span class="p">(</span><span class="o">*</span><span class="n">it</span><span class="p">)]</span> <span class="o">==</span> <span class="mi">4</span><span class="p">)</span> <span class="o">?</span>  <span class="s">"ACGT"</span><span class="p">[</span><span class="n">rand</span><span class="p">()</span><span class="o">%</span><span class="mi">4</span><span class="p">]</span> <span class="o">:</span> <span class="o">*</span><span class="n">it</span><span class="p">);</span>

  <span class="n">v</span><span class="p">.</span><span class="n">count_kmers</span><span class="p">(</span><span class="n">chrom</span><span class="p">);</span>
  <span class="n">chrom</span><span class="p">.</span><span class="n">clear</span><span class="p">();</span>
<span class="p">}</span>
</pre></table></code></div></div><p>and the <code class="language-plaintext highlighter-rouge">encode_char</code> is a static look-up conversion between ASCII characters and two-bit representations. We set 0 for the characters <code class="language-plaintext highlighter-rouge">a</code> and <code class="language-plaintext highlighter-rouge">A</code>, 1 for <code class="language-plaintext highlighter-rouge">c</code> and <code class="language-plaintext highlighter-rouge">C</code>, 2 for <code class="language-plaintext highlighter-rouge">g</code> and <code class="language-plaintext highlighter-rouge">G</code> and 3 for <code class="language-plaintext highlighter-rouge">t</code> and <code class="language-plaintext highlighter-rouge">T</code>. The rest we set to 4, and if we read a character that is a 4 in our look-up, we replace with a nucleotide:</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
</pre><td class="rouge-code"><pre><span class="k">static</span> <span class="k">const</span> <span class="kt">uint8_t</span> <span class="n">encode_char</span><span class="p">[</span><span class="mi">256</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//4</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//17</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//33</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//49</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//@,A-O</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//P-Z</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//`,a-o</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="c1">//p-z</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span>
<span class="p">};</span>
</pre></table></code></div></div><p>Finally, our <code class="language-plaintext highlighter-rouge">main</code> function takes two arguments, the first is a “file of files”, which is a two-column file showing the species name in the first column and the path to its FASTA genome in the second. For example, this is a file called <code class="language-plaintext highlighter-rouge">genomes.txt</code>:</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
</pre><td class="rouge-code"><pre><span class="nv">$ </span><span class="nb">cat </span>genomes.txt
chinese_hamster genomes/criGriChoV2.fa
<span class="nb">cat     </span>genomes/felCat9.fa
gorilla genomes/gorGor6.fa
human   genomes/hg38.fa
mouse   genomes/mm39.fa
alpaca  genomes/vicPac2.fa
whale genomes/balAcu1.fa
</pre></table></code></div></div><p>And our <code class="language-plaintext highlighter-rouge">main</code> function simply reads this file and calls the functions we created. When done, our output will be a table with $4^{12}$ rows and $n$ columns, where $n$ is the number of species. The element in row <code class="language-plaintext highlighter-rouge">i</code> and column <code class="language-plaintext highlighter-rouge">j</code> is the frequency of $k$-mer <code class="language-plaintext highlighter-rouge">i</code> (as its binary representation) in species <code class="language-plaintext highlighter-rouge">j</code>. This is a data frame that we can read into R:</p><div class="language-cpp highlighter-rouge"><div class="code-header"> <span data-label-text="Cpp"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
</pre><td class="rouge-code"><pre><span class="kt">int</span>
<span class="nf">main</span><span class="p">(</span><span class="kt">int</span> <span class="n">argc</span><span class="p">,</span> <span class="k">const</span> <span class="kt">char</span> <span class="o">**</span><span class="n">argv</span><span class="p">)</span> <span class="p">{</span>
  <span class="k">if</span> <span class="p">(</span><span class="n">argc</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s">"usage: ./phylo &lt;input-species.txt&gt;"</span> <span class="o">&lt;&lt;</span> <span class="n">endl</span><span class="p">;</span>
    <span class="k">return</span> <span class="mi">0</span><span class="p">;</span>
  <span class="p">}</span>

  <span class="c1">// ensures the k-mer size used fits in a 32-bit number</span>
  <span class="k">static_assert</span><span class="p">(</span><span class="n">KmerStats</span><span class="o">::</span><span class="n">kmer_size</span> <span class="o">&lt;=</span> <span class="mi">16</span><span class="p">);</span>

  <span class="n">vector</span><span class="o">&lt;</span><span class="n">string</span><span class="o">&gt;</span> <span class="n">species</span><span class="p">;</span>
  <span class="n">vector</span><span class="o">&lt;</span><span class="n">string</span><span class="o">&gt;</span> <span class="n">files</span><span class="p">;</span>

  <span class="n">string</span> <span class="n">tmp1</span><span class="p">,</span> <span class="n">tmp2</span><span class="p">;</span>
  <span class="n">ifstream</span> <span class="n">in</span><span class="p">(</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]);</span>
  <span class="k">while</span> <span class="p">(</span><span class="n">in</span> <span class="o">&gt;&gt;</span> <span class="n">tmp1</span> <span class="o">&gt;&gt;</span> <span class="n">tmp2</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">species</span><span class="p">.</span><span class="n">push_back</span><span class="p">(</span><span class="n">tmp1</span><span class="p">);</span>
    <span class="n">files</span><span class="p">.</span><span class="n">push_back</span><span class="p">(</span><span class="n">tmp2</span><span class="p">);</span>
  <span class="p">}</span>
  <span class="n">in</span><span class="p">.</span><span class="n">close</span><span class="p">();</span>

  <span class="c1">// print the headers: the species names, separated by tabs</span>
  <span class="n">copy</span><span class="p">(</span><span class="n">begin</span><span class="p">(</span><span class="n">species</span><span class="p">),</span> <span class="n">end</span><span class="p">(</span><span class="n">species</span><span class="p">),</span> <span class="n">std</span><span class="o">::</span><span class="n">ostream_iterator</span><span class="o">&lt;</span><span class="n">string</span><span class="o">&gt;</span><span class="p">(</span><span class="n">cout</span><span class="p">,</span> <span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">));</span>
  <span class="n">cout</span> <span class="o">&lt;&lt;</span> <span class="s">"</span><span class="se">\n</span><span class="s">"</span><span class="p">;</span>

  <span class="n">vector</span><span class="o">&lt;</span><span class="n">KmerStats</span><span class="o">&gt;</span> <span class="n">v</span><span class="p">(</span><span class="n">species</span><span class="p">.</span><span class="n">size</span><span class="p">());</span>

  <span class="n">omp_set_num_threads</span><span class="p">(</span><span class="mi">8</span><span class="p">);</span> <span class="c1">// comment if OpenMP not used</span>
<span class="cp">#pragma omp parallel for
</span>  <span class="k">for</span> <span class="p">(</span><span class="kt">size_t</span> <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="n">species</span><span class="p">.</span><span class="n">size</span><span class="p">();</span> <span class="o">++</span><span class="n">i</span><span class="p">)</span> <span class="p">{</span>
<span class="cp">#ifdef VERBOSE
</span>
<span class="cp">#pragma omp critical
</span>    <span class="p">{</span>
      <span class="n">cerr</span> <span class="o">&lt;&lt;</span> <span class="s">"processing "</span> <span class="o">&lt;&lt;</span> <span class="n">species</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">&lt;&lt;</span> <span class="s">"...</span><span class="se">\n</span><span class="s">"</span><span class="p">;</span>
    <span class="p">}</span>
<span class="cp">#endif
</span>    <span class="n">process_species</span><span class="p">(</span><span class="n">files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">v</span><span class="p">[</span><span class="n">i</span><span class="p">]);</span>
  <span class="p">}</span>

<span class="cp">#ifdef VERBOSE
</span>  <span class="n">cerr</span> <span class="o">&lt;&lt;</span> <span class="s">"writing output</span><span class="se">\n</span><span class="s">"</span><span class="p">;</span>
<span class="cp">#endif
</span>  <span class="k">const</span> <span class="kt">size_t</span> <span class="n">num_species</span> <span class="o">=</span> <span class="n">species</span><span class="p">.</span><span class="n">size</span><span class="p">();</span>
  <span class="k">for</span> <span class="p">(</span><span class="kt">size_t</span> <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="n">KmerStats</span><span class="o">::</span><span class="n">num_kmers</span><span class="p">;</span> <span class="o">++</span><span class="n">i</span><span class="p">)</span> <span class="p">{</span>
    <span class="k">for</span> <span class="p">(</span><span class="kt">size_t</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">num_species</span><span class="p">;</span> <span class="o">++</span><span class="n">j</span><span class="p">)</span>
      <span class="n">printf</span><span class="p">(</span><span class="s">"%d</span><span class="se">\t</span><span class="s">"</span><span class="p">,</span> <span class="n">v</span><span class="p">[</span><span class="n">j</span><span class="p">].</span><span class="n">kmer_count</span><span class="p">[</span><span class="n">i</span><span class="p">]);</span>
    <span class="n">printf</span><span class="p">(</span><span class="s">"</span><span class="se">\n</span><span class="s">"</span><span class="p">);</span>
  <span class="p">}</span>
  <span class="k">return</span> <span class="n">EXIT_SUCCESS</span><span class="p">;</span>
<span class="p">}</span>
</pre></table></code></div></div><h2 id="compiling-the-code"><span class="mr-2">Compiling the code</span><a href="#compiling-the-code" class="anchor text-muted"><i class="fas fa-hashtag"></i></a></h2><p>We will create a <code class="language-plaintext highlighter-rouge">Makefile</code> to compile the program. If you do not have OpenMP you can remove the <code class="language-plaintext highlighter-rouge">-fopenmp</code> flag below and the code will run single-thread. It will use less memory and more time, but it should still finish in a few minutes.</p><div class="language-make highlighter-rouge"><div class="code-header"> <span data-label-text="Makefile"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
</pre><td class="rouge-code"><pre><span class="nl">all </span><span class="o">:</span> <span class="nf">phylo</span>

<span class="nl">phylo</span><span class="o">:</span> <span class="nf">src/phylo.cpp</span>
	g++ <span class="nt">-O3</span> <span class="nt">-Wall</span> <span class="nt">-std</span><span class="o">=</span>c++11 <span class="nt">-o</span> phylo src/phylo.cpp <span class="nt">-fopenmp</span>

<span class="nl">clean</span><span class="o">:</span>
	<span class="nb">rm </span>phylo
</pre></table></code></div></div><p>We can compile by simply running</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
</pre><td class="rouge-code"><pre>make all
</pre></table></code></div></div><p>and run the program by writing</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
</pre><td class="rouge-code"><pre>./phylo genomes.txt <span class="o">&gt;</span>kmer-counts.tsv
</pre></table></code></div></div><h2 id="profiling-the-code"><span class="mr-2">Profiling the code</span><a href="#profiling-the-code" class="anchor text-muted"><i class="fas fa-hashtag"></i></a></h2><p>We can further profile using <code class="language-plaintext highlighter-rouge">/usr/bin/time</code> (GNU time) to see how much time and memory it takes. Using 8 cores we get the following:</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
</pre><td class="rouge-code"><pre><span class="nv">$ </span>/usr/bin/time <span class="nt">-v</span> ./phylo genomes.txt <span class="o">&gt;</span>kmer-counts.tsv
processing chinese_hamster...
processing mouse...
processing whale...
processing gorilla...
processing cat...
processing alpaca...
processing human...
writing output
        Command being timed: <span class="s2">"./phylo genomes.txt"</span>
        User <span class="nb">time</span> <span class="o">(</span>seconds<span class="o">)</span>: 525.60
        System <span class="nb">time</span> <span class="o">(</span>seconds<span class="o">)</span>: 17.25
        Percent of CPU this job got: 524%
        Elapsed <span class="o">(</span>wall clock<span class="o">)</span> <span class="nb">time</span> <span class="o">(</span>h:mm:ss or m:ss<span class="o">)</span>: 1:43.54
        Average shared text size <span class="o">(</span>kbytes<span class="o">)</span>: 0
        Average unshared data size <span class="o">(</span>kbytes<span class="o">)</span>: 0
        Average stack size <span class="o">(</span>kbytes<span class="o">)</span>: 0
        Average total size <span class="o">(</span>kbytes<span class="o">)</span>: 0
        Maximum resident <span class="nb">set </span>size <span class="o">(</span>kbytes<span class="o">)</span>: 1653768
        Average resident <span class="nb">set </span>size <span class="o">(</span>kbytes<span class="o">)</span>: 0
        Major <span class="o">(</span>requiring I/O<span class="o">)</span> page faults: 0
        Minor <span class="o">(</span>reclaiming a frame<span class="o">)</span> page faults: 412911
        Voluntary context switches: 659062
        Involuntary context switches: 91400
        Swaps: 0
        File system inputs: 0
        File system outputs: 805360
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size <span class="o">(</span>bytes<span class="o">)</span>: 4096
        Exit status: 0
</pre></table></code></div></div><p>So we used a little under 2 minutes and 1.6 GB to create our table. Note that the high memory use is because we used OpenMP to count the k-mers of all 7 species in parallel, so all <code class="language-plaintext highlighter-rouge">KmerStats</code> objects were loaded. Each <code class="language-plaintext highlighter-rouge">KmerStats</code> object allocates a vector of size $4^{12}$ with 32 bits, so each takes 64 MB. Then each thread uses an additional 250 MB as pre-allocation for the chromosomes.</p><p>We can further profile the code using <a href="https://www.man7.org/linux/man-pages/man1/perf.1.html">perf</a>.</p><div class="language-bash highlighter-rouge"><div class="code-header"> <span data-label-text="Shell"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
</pre><td class="rouge-code"><pre>perf record <span class="nt">-v</span> ./phylo genomes.txt <span class="o">&gt;</span>kmer-counts.tsv
perf report
</pre></table></code></div></div><h1 id="the-r-code">The R code</h1><p>For the final part, we will construct a hierarchical clustering based on the $n \times n$ distance matrix, defined as the sum of squares of the differences between $k$-mer frequencies for any two species. We will load the output of the C++ program into a matrix <code class="language-plaintext highlighter-rouge">x</code>, calculate the pairwise distances using the <code class="language-plaintext highlighter-rouge">dist</code> function (we need to transpose because <code class="language-plaintext highlighter-rouge">dist</code> is between all pairs of rows), then use <code class="language-plaintext highlighter-rouge">heatmap</code> and <code class="language-plaintext highlighter-rouge">hclust</code> to make plots.</p><div class="language-r highlighter-rouge"><div class="code-header"> <span data-label-text="R"><i class="fas fa-code small"></i></span> <button aria-label="copy" data-title-succeed="Copied!"><i class="far fa-clipboard"></i></button></div><div class="highlight"><code><table class="rouge-table"><tbody><tr><td class="rouge-gutter gl"><pre class="lineno">1
2
3
4
</pre><td class="rouge-code"><pre><span class="o">&gt;</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="s1">'kmer-counts.tsv'</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="o">&gt;</span><span class="w"> </span><span class="n">the.dist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dist</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">x</span><span class="p">))</span><span class="w"> </span><span class="c1"># distance between all pairs of species</span><span class="w">
</span><span class="o">&gt;</span><span class="w"> </span><span class="n">heatmap</span><span class="p">(</span><span class="n">the.dist</span><span class="p">)</span><span class="w"> </span><span class="c1"># heatmap of distance matrix</span><span class="w">
</span><span class="o">&gt;</span><span class="w"> </span><span class="n">plot</span><span class="p">(</span><span class="n">hclust</span><span class="p">(</span><span class="n">the.dist</span><span class="p">),</span><span class="w"> </span><span class="n">hang</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-1</span><span class="p">)</span><span class="w"> </span><span class="c1"># phylogenetic tree</span><span class="w">
</span></pre></table></code></div></div><p>This is the heatmap result: <img data-src="https://i.ibb.co/JpqbJ8R/heatmap.png" alt="heatmap of distances" data-proofer-ignore></p><p>And this is the resulting tree</p><p><img data-src="https://i.ibb.co/Syhv1hb/phylo.png" alt="our resulting phylogenetic tree" data-proofer-ignore></p><p>If you compare where the species lie in the <a href="https://genome.ucsc.edu/cgi-bin/hgGateway">UCSC genome browser tree</a>, you will see that our tree is consistent with theirs. Human and gorilla cluster together, as do mouse and hamster. Whale, cat and alpaca cluster together, with whale closer to alpaca. Hooray!</p><h1 id="limitations-of-the-method">Limitations of the method</h1><p>This algorithm is admittedly naive. First, it treats the $k$-mer frequencies in isolation. We have not accounted for co-occurences of $k$-mers. Consider the $k$-mer <code class="language-plaintext highlighter-rouge">AAAAAAAAAAAA</code> and three genomes. In genomes A and B, we have a large run of 1000 consecutive A, resulting in 989 occurrences of this $k$-mer. Then, in genome C, the sequence <code class="language-plaintext highlighter-rouge">AAAAAAAAAAAA</code> appears uniformly at random in the middle of larger more complex sequences. The contribution of <code class="language-plaintext highlighter-rouge">AAAAAAAAAAAA</code> is identical in all three genomes, but the correlation of $k$-mers in A and B implies that they are more similar. We can incorporate co-occurrences of sequences by modeling them as Markov chains, and comparing the Markov chain parameters instead of the $k$-mers directly. There is certainly a lot to improvement avenues to explore, but our idea is a start.</p><h1 id="follow-up-questions">Follow-up questions</h1><p>We can use a similar approach to guess, given a sequencing dataset (e.g. Illumina, PacBio or Oxford Nanopore), which species it most likely comes from. Can you think about how to do this? What adjustments are required in this method?</p></div><div class="post-tail-wrapper text-muted"><div class="post-meta mb-3"> <i class="far fa-folder-open fa-fw mr-1"></i> <a href='/categories/c/'>C++</a>, <a href='/categories/phylogenetics/'>phylogenetics</a></div><div class="post-tail-bottom d-flex justify-content-between align-items-center mt-3 pt-5 pb-2"><div class="license-wrapper"> This post is licensed under <a href="https://creativecommons.org/licenses/by/4.0/"> CC BY 4.0 </a> by the author.</div><div class="share-wrapper"> <span class="share-label text-muted mr-1">Share</span> <span class="share-icons"> <a href="https://t.me/share/url?url=%2Fposts%2Fpossibly-the-most-naive-phylogenetic-reconstruction-algorithm%2F&text=The+most+naive+phylogenetic+reconstruction+algorithm+-+Guilherme+Sena" data-toggle="tooltip" data-placement="top" title="Telegram" target="_blank" rel="noopener" aria-label="Telegram"> <i class="fa-fw fab fa-telegram"></i> </a> <i id="copy-link" class="fa-fw fas fa-link small" data-toggle="tooltip" data-placement="top" title="Copy link" data-title-succeed="Link copied successfully!"> </i> </span></div></div></div></div></div><div id="panel-wrapper" class="col-xl-3 pl-2 text-muted"><div class="access"><div id="access-lastmod" class="post"><div class="panel-heading">Recently Updated</div><ul class="post-content pl-0 pb-1 ml-1 mt-2"><li><a href="/posts/possibly-the-most-naive-phylogenetic-reconstruction-algorithm/">The most naive phylogenetic reconstruction algorithm</a><li><a href="/posts/setting-up-a-blog/">Setting up a blog</a></ul></div></div><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@1.0.1/dist/bootstrap-toc.min.js"></script><div id="toc-wrapper" class="pl-0 pr-4 mb-5"><div class="panel-heading pl-3 pt-2 mb-2">Contents</div><nav id="toc" data-toggle="toc"></nav></div></div></div><div class="row"><div id="tail-wrapper" class="col-12 col-lg-11 col-xl-9 pl-3 pr-3 pr-xl-4"><div id="related-posts" class="mt-5 mb-2 mb-sm-4"><h3 class="pt-2 mt-1 mb-4 ml-1" data-toc-skip>Further Reading</h3><div class="card-deck mb-4"><div class="card"> <a href="/posts/wavefront-pt-1/"><div class="card-body"> <em class="small" data-ts="1658448000" data-df="ll" > Jul 22, 2022 </em><h3 class="pt-0 mt-1 mb-3" data-toc-skip>Wavefront alignments pt 1 : The Myers edit distance algorithm</h3><div class="text-muted small"><p> This is the first part of a two-part post that explains the wavefront alignment algorithm, a 2020 paper that finds the alignment between two sequences $A$ and $B$ in $O((|A| + |B|)s)$, where $s$ is...</p></div></div></a></div><div class="card"> <a href="/posts/setting-up-a-blog/"><div class="card-body"> <em class="small" data-ts="1576022400" data-df="ll" > Dec 11, 2019 </em><h3 class="pt-0 mt-1 mb-3" data-toc-skip>Setting up a blog</h3><div class="text-muted small"><p> I am configuring this blog if I want to communicate, write or document things using markdown and figures. This post is to test if I can write some code with syntax highlight and some math. Here’s ...</p></div></div></a></div></div></div><div class="post-navigation d-flex justify-content-between"> <a href="/posts/setting-up-a-blog/" class="btn btn-outline-primary" prompt="Older"><p>Setting up a blog</p></a> <a href="/posts/wavefront-pt-1/" class="btn btn-outline-primary" prompt="Newer"><p>Wavefront alignments pt 1 : The Myers edit distance algorithm</p></a></div></div></div><footer class="row pl-3 pr-3"><div class="col-12 d-flex justify-content-between align-items-center text-muted pl-0 pr-0"><div class="footer-left"><p class="mb-0"> © 2022 <a href="https://twitter.com/senacompbio">Guilherme Sena</a>. <span data-toggle="tooltip" data-placement="top" title="Except where otherwise noted, the blog posts on this site are licensed under the Creative Commons Attribution 4.0 International (CC BY 4.0) License by the author.">Some rights reserved.</span></p></div><div class="footer-right"><p class="mb-0"> Powered by <a href="https://jekyllrb.com" target="_blank" rel="noopener">Jekyll</a> with <a href="https://github.com/cotes2020/jekyll-theme-chirpy" target="_blank" rel="noopener">Chirpy</a> theme.</p></div></div></footer></div><div id="search-result-wrapper" class="d-flex justify-content-center unloaded"><div class="col-12 col-sm-11 post-content"><div id="search-hints"></div><div id="search-results" class="d-flex flex-wrap justify-content-center text-muted mt-3"></div></div></div></div><div id="mask"></div><a id="back-to-top" href="#" aria-label="back-to-top" class="btn btn-lg btn-box-shadow" role="button"> <i class="fas fa-angle-up"></i> </a><div id="notification" class="toast" role="alert" aria-live="assertive" aria-atomic="true" data-animation="true" data-autohide="false"><div class="toast-header"> <button type="button" class="ml-2 ml-auto close" data-dismiss="toast" aria-label="Close"> <span aria-hidden="true">&times;</span> </button></div><div class="toast-body text-center pt-0"><p class="pl-2 pr-2 mb-3">A new version of content is available.</p><button type="button" class="btn btn-primary" aria-label="Update"> Update </button></div></div><script src="https://cdn.jsdelivr.net/npm/simple-jekyll-search@1.10.0/dest/simple-jekyll-search.min.js"></script> <script> SimpleJekyllSearch({ searchInput: document.getElementById('search-input'), resultsContainer: document.getElementById('search-results'), json: '/assets/js/data/search.json', searchResultTemplate: '<div class="pl-1 pr-1 pl-sm-2 pr-sm-2 pl-lg-4 pr-lg-4 pl-xl-0 pr-xl-0"> <a href="{url}">{title}</a><div class="post-meta d-flex flex-column flex-sm-row text-muted mt-1 mb-1"> {categories} {tags}</div><p>{snippet}</p></div>', noResultsText: '<p class="mt-5">Oops! No results found.</p>', templateMiddleware: function(prop, value, template) { if (prop === 'categories') { if (value === '') { return `${value}`; } else { return `<div class="mr-sm-4"><i class="far fa-folder fa-fw"></i>${value}</div>`; } } if (prop === 'tags') { if (value === '') { return `${value}`; } else { return `<div><i class="fa fa-tag fa-fw"></i>${value}</div>`; } } } }); </script> <script src="https://cdn.jsdelivr.net/combine/npm/magnific-popup@1/dist/jquery.magnific-popup.min.js,npm/lozad/dist/lozad.min.js,npm/clipboard@2/dist/clipboard.min.js"></script> <script src="https://cdn.jsdelivr.net/combine/npm/dayjs@1/dayjs.min.js,npm/dayjs@1/locale/en.min.js,npm/dayjs@1/plugin/relativeTime.min.js,npm/dayjs@1/plugin/localizedFormat.min.js"></script> <script defer src="/assets/js/dist/post.min.js"></script> <script> /* see: <https://docs.mathjax.org/en/latest/options/input/tex.html#tex-options> */ MathJax = { tex: { inlineMath: [ /* start/end delimiter pairs for in-line math */ ['$','$'], ['\\(','\\)'] ], displayMath: [ /* start/end delimiter pairs for display math */ ['$$', '$$'], ['\\[', '\\]'] ] } }; </script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"> </script> <script src="https://cdn.jsdelivr.net/npm/bootstrap@4/dist/js/bootstrap.bundle.min.js"></script> <script defer src="/app.js"></script>
